packages/flutter_tools/lib/src/daemon.dart

```diff
@@ -3,108 +3,192 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
 
 import 'base/common.dart';
 import 'base/io.dart';
 import 'base/logger.dart';
 import 'base/utils.dart';
 import 'convert.dart';
 
-/// Parse binary streams in the JSON RPC format understood by the daemon, and
-/// convert it into a stream of JSON RPC messages.
-Stream<Map<String, Object?>> _convertInputStream(Stream<List<int>> inputStream) {
-  return utf8.decoder.bind(inputStream)
-  .transform<String>(const LineSplitter())
-  .where((String line) => line.startsWith('[{') && line.endsWith('}]'))
-  .map<Map<String, Object?>?>((String line) {
-    line = line.substring(1, line.length - 1);
-    return castStringKeyedMap(json.decode(line));
-  })
-  .where((Map<String, Object?>? entry) => entry != null)
-  .cast<Map<String, Object?>>();
+/// A single message passed through the [DaemonConnection].
+class DaemonMessage {
+  DaemonMessage(this.data, [this.binary]);
+
+  /// Content of the JSON message in the message.
+  final Map<String, Object?> data;
+
+  /// Stream of the binary content of the message.
+  ///
+  /// Must be listened to if binary data is present.
+  final Stream<List<int>>? binary;
 }
 
-/// A stream that a [DaemonConnection] uses to communicate with each other.
-abstract class DaemonStreams {
-  /// Stream that contains input to the [DaemonConnection].
-  Stream<Map<String, Object?>> get inputStream;
+/// Data of an event passed through the [DaemonConnection].
+class DaemonEventData {
+  DaemonEventData(this.eventName, this.data, [this.binary]);
 
-  /// Outputs a message through the connection.
-  void send(Map<String, Object?> message);
+  /// The name of the event.
+  final String eventName;
 
-  /// Cleans up any resources used.
-  Future<void> dispose() async { }
+  /// The data of the event.
+  final Object? data;
+
+  /// Stream of the binary content of the event.
+  ///
+  /// Must be listened to if binary data is present.
+  final Stream<List<int>>? binary;
 }
 
-/// A [DaemonStream] that uses stdin and stdout as the underlying streams.
-class StdioDaemonStreams extends DaemonStreams {
-  StdioDaemonStreams(Stdio stdio) :
-    _stdio = stdio,
-    inputStream = _convertInputStream(stdio.stdin);
+const String _binaryLengthKey = '_binaryLength';
 
-  final Stdio _stdio;
+enum _InputStreamParseState {
+  JSON,
+  BINARY,
```

Comment:
Lower camel https://dart.dev/guides/language/effective-dart/style#prefer-using-lowercamelcase-for-constant-names
```suggestion
  json,
  binary,
```
>  Note: We initially used Java’s SCREAMING_CAPS style for constants. We changed for a few reasons:
> SCREAMING_CAPS looks bad for many cases, particularly enum values for things like CSS colors.
> Constants are often changed to final non-const variables, which would necessitate a name change.
> The values property automatically defined on an enum type is const and lowercase.

Reply:
Done, changed to lower case